### PR TITLE
Frontend coincontrol feedback

### PR DESCRIPTION
**Pull request by originalblend337**
Will need to add a guide entry separately if we want to elaborate on how to use coin control.

### CODE REVIEW

File: frontends/web/src/decorators/translate.ts [hunk 1]
```diff
@@ -32,14 +32,14 @@ interface InterpolationValues {
 /**
  * This interface describes the default value to return if no translation was found.
  */
-// interface DefaultValue {
-//     defaultValue: any;
-// }
+interface DefaultValue {
```
Comment (benma) [inline frontends/web/src/decorators/translate.ts]: What does this commit do, why is this needed?

**Comment by benma** [comment]
CI: 

> i18n files malformatted. Fix with: make locize-fix

File: frontends/web/src/routes/account/send/send.tsx [hunk 2]
```diff
@@ -16,33 +16,93 @@
...
+    deviceIDs: string[];
+}
+
+interface AccountProps {
+    code: string;
+    coinCode: string;
+}
+
+interface ProposedAmountProps {
+    amount: string | null;
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: I would quite strongly prefer to use `amount?: string` (i.e. with the implicit `undefined`). In case you agree and just tried to model how the component was implemented before using TypeScript in this commit, then it's ok for me to leave this for a future commit.

Comment (KasparEtter) [reply frontends/web/src/routes/account/send/send.tsx]: And on a different note: Shouldn't the logic be that *if* an `Amount` is provided, then the amount (and unit) field has to be set? It seems to me that you are making too many things optional here.

File: frontends/web/src/routes/account/send/send.tsx [hunk 3]
```diff
@@ -16,33 +16,93 @@
...
+    unit: string | null;
+    conversions: ConversionsProps | null;
+}
+
+interface ConversionsProps {
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Remove `Props`.

File: frontends/web/src/routes/account/send/send.tsx [hunk 4]
```diff
@@ -16,33 +16,93 @@
...
+    [key: string]: string | null;
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Do we really want to allow the object to be `{ "value": null }`? All values are already optional in the sense that you can simply not list them at all. Moreover, if I understand this correctly (and a more specific type would really help in this regard), you could use the following type for the `key`: https://github.com/digitalbitbox/bitbox-wallet-app/blob/06536915b08a3added0d7d3106e03bedc2a22225/frontends/web/src/components/rates/rates.tsx#L32

File: frontends/web/src/routes/account/send/send.tsx [hunk 5]
```diff
@@ -16,33 +16,93 @@
...
+}
+
+interface SignProgressProps {
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: No `Props`. And do I interpret this correctly that `steps` is `totalSteps` and `step` is `currentStep`? Again, I guess you just modelled the existing implementation so leaving this for a separate commit/PR is fine.

File: frontends/web/src/routes/account/send/send.tsx [hunk 6]
```diff
@@ -16,33 +16,93 @@
...
+    addressError: string | null;
+    amountError: string | null;
+    dataError: string | null;
+    paired: boolean | null;
+    noMobileChannelError: boolean;
+    signProgress: SignProgressProps | null;
+    signConfirm: boolean | null;
+    coinControl: boolean;
+    activeCoinControl: boolean;
+}
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: I did not study (yet) whether the right fields are optional but as above I would prefer to make them really optional (with the `?`) so that you don't have to list them all if you declare the object.

File: frontends/web/src/routes/account/send/send.tsx [hunk 7]
```diff
@@ -16,33 +16,93 @@
...
-export default class Send extends Component {
+class Send extends Component<Props, State> {
+    private utxos!: Element;
+    private selectedUTXOs!: SelectedUTXOProps;
+    private unsubscribe!: () => void;
 
     constructor(props) {
         super(props);
-
+        this.selectedUTXOs = {};
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: I guess you could just declare it this way which would also make the `!` unnecessary: `private selectedUTXOs: SelectedUTXOProps = {};`.

File: frontends/web/src/routes/account/send/send.tsx [hunk 8]
```diff
@@ -65,24 +125,24 @@ export default class Send extends Component {
...
             activeCoinControl: false,
+            data: null,
         };
-        this.selectedUTXOs = [];
     }
 
-    coinSupportsCoinControl = () => {
-        const account = this.getAccount();
+    private coinSupportsCoinControl = () => {
+        const { account } = this.state;
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Is this preferable to `const account = this.state.account;`? And since you use the account twice at most, I feel that the local const is no longer really necessary.

File: frontends/web/src/routes/account/send/send.tsx [hunk 9]
```diff
@@ -154,13 +213,15 @@ export default class Send extends Component {
                     data: null,
                 });
                 if (this.utxos) {
+                    // @ts-ignore
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: I would prefer if you cast the `getWrappedInstance()` (Which I guess is necessary? Or what is its type?). After you made `clear()` public, shouldn't the following (or at least something like it) work?

`(this.utxos.getWrappedInstance() as UTXOs).clear();`

This would have the advantage that TypeScript would help us detect if the method changed (its name or arguments) or is no longer there.

File: frontends/web/src/routes/account/send/send.tsx [hunk 10]
```diff
@@ -298,89 +364,101 @@ export default class Send extends Component {
...
-        activeCoinControl,
-    }) {
-        const account = this.getAccount();
-        if (!account) return null;
+    private setUTXOsRef = (ref: Element) => {
+        this.utxos = ref;
+    }
+
+    public render(
+        { t, code }: Props,
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: I think `RenderableProps<Props>` is always a good idea.

File: frontends/web/src/routes/account/send/send.tsx [hunk 11]
```diff
@@ -298,89 +364,101 @@ export default class Send extends Component {
...
+                <br />
                 {t('send.signprogress.label')}: {signProgress.step}/{signProgress.steps}
             </span>
         ) : null;
         return (
             <div class="contentWithGuide">
                 <div class="container">
                     <Status type="warning">
-                        {paired === false && t('warning.sendPairing')}
+                        {paired === false ? t('warning.sendPairing') : null}
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: As mentioned before, upgrading Preact should make this change unnecessary.

File: frontends/web/src/routes/account/send/send.tsx [hunk 12]
```diff
@@ -298,89 +364,101 @@ export default class Send extends Component {
...
                     </Status>
                     <Header title={<h2>{t('send.title')}</h2>}>
-                        <Balance
-                            balance={balance} />
+                    <Balance balance={balance} />
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: If the `Balance` is in the `Header`, it should be indented, shouldn't it?

File: frontends/web/src/routes/account/send/send.tsx [hunk 13]
```diff
@@ -392,13 +470,14 @@ export default class Send extends Component {
                     <div class="innerContainer scrollableContainer">
                         <div class="content padded">
                             {
-                                coinControl && (
+                                coinControl ? (
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Dito regarding Preact.

File: frontends/web/src/routes/account/send/send.tsx [hunk 14]
```diff
@@ -410,11 +489,13 @@ export default class Send extends Component {
                                     value={recipientAddress}
                                     autofocus
                                 />
-                                { debug && (
-                                    <span id="sendToSelf" className={style.action} onClick={this.sendToSelf}>
-                                        Send to self
-                                    </span>
-                                ) }
+                                {
+                                    debug ? (
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Dito regarding Preact.

File: frontends/web/src/routes/account/send/send.tsx [hunk 15]
```diff
@@ -466,19 +547,24 @@ export default class Send extends Component {
                                     />
                                     */}
                                 </div>
-                                {feeTarget && <p class={style.feeDescription}>{t('send.feeTarget.description.' + feeTarget)}</p>}
+                                {
+                                    feeTarget ? (
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Dito regarding Preact.

File: frontends/web/src/routes/account/send/send.tsx [hunk 16]
```diff
@@ -466,19 +547,24 @@ export default class Send extends Component {
...
-                                    <Input
-                                        label={t('send.data.label')}
-                                        placeholder={t('send.data.placeholder')}
-                                        id="data"
-                                        error={dataError}
-                                        onInput={this.handleFormChange}
-                                        value={data}
-                                    />
-                                </div>
+                                (account.coinCode === 'eth' || account.coinCode === 'teth' || account.coinCode === 'reth') ? (
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Dito regarding Preact.

Comment (KasparEtter) [reply frontends/web/src/routes/account/send/send.tsx]: Also a few more occurrences of this below which I will no longer comment.

File: frontends/web/src/locales/en/app.json [hunk 17]
```diff
@@ -654,6 +654,7 @@
     },
     "button": "Sign and Send",
     "coincontrol": {
+      "title": "Send from output",
```
Comment (KasparEtter) [inline frontends/web/src/locales/en/app.json]: I haven't studied where this is used but if it says `title` it should probably be in title-case (with a capital `O` in output).

File: frontends/web/src/decorators/translate.ts (hunk 1 above)
Comment (KasparEtter) [reply frontends/web/src/decorators/translate.ts]: Ah, then please ignore (parts of) my comments before.

File: frontends/web/src/routes/device/deviceswitch.tsx [hunk 18]
```diff
@@ -24,6 +24,7 @@ export interface Devices {
 }
 
 interface Props {
+    key?: any;
```
Comment (originalblend337) [inline frontends/web/src/routes/device/deviceswitch.tsx]: @KasparEtter @benma  -- after updating preact to the latest, TS has been complaining that it cannot assign `devices` to the property `key` (occurs in `app.tsx`), do you know a better work around for this than defining it as `any`?

Comment (KasparEtter) [reply frontends/web/src/routes/device/deviceswitch.tsx]: Can you post the exact error message here? Or otherwise tell me what I need to look for resp. change if I check out your branch?

Comment (originalblend337) [reply frontends/web/src/routes/device/deviceswitch.tsx]: First error I get is
```
✖ ERROR [at-loader] ./src/app.tsx:181:29
    TS2326: Types of property 'key' are incompatible.
  Type 'Devices' is not assignable to type 'string | number | undefined'.
    Type 'Devices' is not assignable to type 'number'.
✖ ERROR [at-loader] ./src/app.tsx:186:29
    TS2326: Types of property 'key' are incompatible.
  Type 'Devices' is not assignable to type 'string | number | undefined'.
    Type 'Devices' is not assignable to type 'number'.
```

and when I try to define key to be `Devices`, which is
```
interface Devices {
    [deviceID: string]: string;
}
```
I get
```
✖ ERROR [at-loader] ./src/app.tsx:181:29
    TS2326: Types of property 'key' are incompatible.
  Type 'Devices' is not assignable to type '(Devices & string) | (Devices & number) | undefined'.
    Type 'Devices' is not assignable to type 'Devices & number'.
      Type 'Devices' is not assignable to type 'number'.
✖ ERROR [at-loader] ./src/app.tsx:186:29
    TS2326: Types of property 'key' are incompatible.
  Type 'Devices' is not assignable to type '(Devices & string) | (Devices & number) | undefined'.
    Type 'Devices' is not assignable to type 'Devices & number'.
      Type 'Devices' is not assignable to type 'number'.
```

File: frontends/web/src/routes/account/send/send.tsx [hunk 19]
```diff
@@ -65,26 +124,26 @@ export default class Send extends Component {
...
         if (this.props.deviceIDs.length > 0) {
-            apiGet('devices/' + this.props.deviceIDs[0] + '/has-mobile-channel').then(mobileChannel => {
+            apiGet('devices/' + this.props.deviceIDs[0] + '/has-mobile-channel').then((mobileChannel: boolean) => {
                 apiGet('devices/' + this.props.deviceIDs[0] + '/info').then(({ pairing }) => {
-                    const account = this.getAccount();
+                    const account = this.state.account;
+                    const paired = mobileChannel && pairing;
                     const noMobileChannelError = pairing && !mobileChannel && account && isBitcoinBased(account.coinCode);
-                    this.setState({ paired: mobileChannel && pairing, noMobileChannelError });
+                    this.setState(prevState => ({ ...prevState, paired, noMobileChannelError }));
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Why do we need the `prevState`? Why not just `this.setState({ paired, noMobileChannelError });`?

File: frontends/web/src/routes/account/send/send.tsx [hunk 20]
```diff
@@ -14,35 +14,94 @@
...
+import { debug } from '../../../utils/env';
+import { apiGet, apiPost } from '../../../utils/request';
+import { apiWebsocket } from '../../../utils/websocket';
+import { isBitcoinBased } from '../utils';
 import FeeTargets from './feetargets';
-import UTXOs from './utxos';
-import approve from '../../../assets/icons/checked.svg';
-import reject from '../../../assets/icons/cancel.svg';
 import * as style from './send.css';
+import { Props as UTXOsProps, SelectedUTXO, UTXOs } from './utxos';
```
Comment (KasparEtter) [inline frontends/web/src/routes/account/send/send.tsx]: Why do you import `Props` (instead of `UTXOsProps `)? It is declared as `export type Props = UTXOsProps & TranslateProps` and we don't need the `t` here.
